src/memory/mmap_bytebuf.rs

```diff
@@ -5,181 +5,121 @@
-            //   of `self`.
-            libc::munmap(self.ptr.as_ptr() as *mut libc::c_void, self.len)
-        };
-
-        // There is no reason for `munmap` to fail to deallocate a private annonymous mapping
-        // allocated by `mmap`.
-        // However, for the cases when it actually fails prefer to fail, in order to not leak
-        // and exhaust the virtual memory.
-        assert_eq!(ret_val, 0, "munmap failed");
+    /// Returns an exclusive slice over the bytes of the virtual memory allocation.
+    pub fn as_slice_mut(&mut self) -> &mut [u8] {
+        // # SAFETY
+        //
+        // See safety proof of the `as_slice` method.
+        // Additionally, it is not possible to obtain two mutable references for the same memory area.
+        unsafe { slice::from_raw_parts_mut(self.allocation.as_mut_ptr(), self.allocation.len()) }
     }
 }
 
+/// A virtually allocated byte buffer.
 pub struct ByteBuf {
-    mmap: Option<Mmap>,
+    /// The underlying virtual memory allocation.
+    mem: VirtualMemory,
+    /// The current size of the used parts of the virtual memory allocation.
+    len: usize,
 }
 
 impl ByteBuf {
+    /// Determines the initial size of the virtual memory allocation.
+    ///
+    /// # Note
+    ///
+    /// In this implementation we won't reallocate the virtually allocated
+    /// buffer and instead simply adjust the `len` field of the `ByteBuf`
+    /// wrapper in order to efficiently grow the virtual memory.
+    const ALLOCATION_SIZE: usize = u32::MAX as usize;
+
+    /// Creates a new byte buffer with the given initial length.
     pub fn new(len: usize) -> Result<Self, String> {
-        let mmap = if len == 0 {
-            None
-        } else {
-            Some(Mmap::new(len)?)
-        };
-        Ok(Self { mmap })
+        if len > isize::max_value() as usize {
+            return Err("`len` should not exceed `isize::max_value()`".into());
+        }
+        let mem = VirtualMemory::new(Self::ALLOCATION_SIZE)?;
+        Ok(Self { mem, len })
     }
 
+    /// Reallocates the virtual memory with the new length in bytes.
     pub fn realloc(&mut self, new_len: usize) -> Result<(), String> {
-        let new_mmap = if new_len == 0 {
-            None
-        } else {
-            let mut new_mmap = Mmap::new(new_len)?;
-            if let Some(cur_mmap) = self.mmap.take() {
-                let src = cur_mmap.as_slice();
-                let dst = new_mmap.as_slice_mut();
-                let amount = src.len().min(dst.len());
-                dst[..amount].copy_from_slice(&src[..amount]);
-            }
-            Some(new_mmap)
-        };
-
-        self.mmap = new_mmap;
+        // This operation is only actually needed in order to make the
+        // Vec-based implementation less inefficient. In the case of a
+        // virtual memory with preallocated 4GB of virtual memory pages
+        // we only need to adjust the `len` field.
```

Comment:
We are in `target_pointer_width = "64"` so usize::BITS=64. Is there anything to stop us from `.realloc(2^32+1)`?

Reply:
no.
we are aware about guard pages with additional page allocation but wanted to keep the implementation simple.

Reply:
Wouldn't `assert!(new_len <= ALLOCATION_SIZE);` keep the impl simple?

Reply:
The `len` should always reflect the current size of the linear memory in wasm semantics. We need to bound check memory access.

Reply:
I mean one can do `byte_buf.realloc(2^32+1)` and then call `byte_buf.as_slice()` which will give a slice that references to a out-of-bound mem